File: src/cpp/src/barrier_pricer.cpp
#include "lsv/barrier_pricer.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>

#include "lsv/brownian_bridge.hpp"

namespace lsv {

BarrierPriceResult price_barrier_paths(
    std::span<const double> spots,
    std::span<const double> interval_variances, const std::size_t paths,
    const std::size_t steps, const double maturity, const double rate,
    const double payout, const BarrierType barrier_type,
    const BarrierPayoff payoff_type, const Monitoring monitoring,
    const double lower_barrier, const double upper_barrier,
    const std::uint64_t seed, const int image_terms) {
  if (paths == 0U || steps == 0U || spots.size() != paths * (steps + 1U) ||
      interval_variances.size() != paths * steps || maturity <= 0.0 ||
      payout < 0.0 || !std::isfinite(rate) || !std::isfinite(payout)) {
    throw std::invalid_argument("invalid barrier pricing inputs");
  }
  const bool uses_lower =
      barrier_type == BarrierType::Lower || barrier_type == BarrierType::Double;
  const bool uses_upper =
      barrier_type == BarrierType::Upper || barrier_type == BarrierType::Double;
  if ((uses_lower && (!(lower_barrier > 0.0) ||
                      lower_barrier >= spots.front())) ||
      (uses_upper && (!(upper_barrier > spots.front()))) ||
      (barrier_type == BarrierType::Double &&
       lower_barrier >= upper_barrier)) {
    throw std::invalid_argument("barriers must strictly bracket initial spot");
  }
  const double log_lower = uses_lower ? std::log(lower_barrier) : 0.0;
  const double log_upper = uses_upper ? std::log(upper_barrier) : 0.0;
  const double discount = std::exp(-rate * maturity);
  double sum = 0.0;
  double squared_sum = 0.0;
  for (std::size_t path = 0; path < paths; ++path) {
    const std::size_t spot_row = path * (steps + 1U);
    const std::size_t variance_row = path * steps;
    double survival = 1.0;
    for (std::size_t step = 0; step < steps; ++step) {
      const double start = spots[spot_row + step];
      const double end = spots[spot_row + step + 1U];
      const double variance = interval_variances[variance_row + step];
      if (!(start > 0.0) || !(end > 0.0) || variance < 0.0 ||
          !std::isfinite(variance)) {
        throw std::invalid_argument("paths and interval variances must be valid");
      }
      const double log_start = std::log(start);
      const double log_end = std::log(end);
      double interval_survival = 1.0;
      if (barrier_type == BarrierType::Lower) {
        interval_survival =
            monitoring == Monitoring::Continuous
                ? lower_bridge_survival(log_start, log_end, log_lower, variance)
                : (end > lower_barrier ? 1.0 : 0.0);
      } else if (barrier_type == BarrierType::Upper) {
        interval_survival =
            monitoring == Monitoring::Continuous
                ? upper_bridge_survival(log_start, log_end, log_upper, variance)
                : (end < upper_barrier ? 1.0 : 0.0);
      } else {
        interval_survival =
            monitoring == Monitoring::Continuous
                ? double_bridge_survival(log_start, log_end, log_lower,
                                         log_upper, variance, image_terms)
                : (end > lower_barrier && end < upper_barrier ? 1.0 : 0.0);
      }
      survival *= interval_survival;
      if (survival == 0.0) {
        break;
      }
    }
    const double probability =
        payoff_type == BarrierPayoff::NoTouch ? survival : 1.0 - survival;
    const double discounted_payoff = discount * payout * probability;
    sum += discounted_payoff;
    squared_sum += discounted_payoff * discounted_payoff;
  }
  const double count = static_cast<double>(paths);
  const double estimate = sum / count;
  const double sample_variance =
      paths > 1U
          ? std::max((squared_sum - count * estimate * estimate) / (count - 1.0),
                     0.0)
          : 0.0;
  const double standard_error = std::sqrt(sample_variance / count);
  constexpr double z95 = 1.959963984540054;
  return {estimate,
          standard_error,
          std::max(0.0, estimate - z95 * standard_error),
          std::min(discount * payout, estimate + z95 * standard_error),
          paths,
          steps,
          seed};
}

}  // namespace lsv

File: src/python/lsv/barrier_pricer.py
"""Typed Python orchestration for C++ barrier and DNT pricing."""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

import lsv_cpp
import numpy as np
from numpy.typing import ArrayLike


class BarrierType(StrEnum):
    """Barrier geometry."""

    LOWER = "lower"
    UPPER = "upper"
    DOUBLE = "double"


class BarrierPayoff(StrEnum):
    """Whether the maturity payout is conditional on touch or survival."""

    TOUCH = "touch"
    NO_TOUCH = "no_touch"


class Monitoring(StrEnum):
    """Barrier monitoring convention."""

    DISCRETE = "discrete"
    CONTINUOUS = "continuous"


@dataclass(frozen=True)
class BarrierPrice:
    """Monte Carlo estimate and 95% normal confidence interval."""

    estimate: float
    standard_error: float
    confidence_interval: tuple[float, float]
    seed: int
    paths: int
    steps: int
    monitoring: Monitoring


def price_barrier(
    spots: ArrayLike,
    variances: ArrayLike,
    maturity: float,
    rate: float,
    payout: float,
    barrier_type: BarrierType | str,
    payoff_type: BarrierPayoff | str,
    monitoring: Monitoring | str,
    *,
    lower_barrier: float | None = None,
    upper_barrier: float | None = None,
    local_variances: ArrayLike | None = None,
    seed: int = 0,
    image_terms: int = 12,
) -> BarrierPrice:
    """Price a maturity-paid barrier claim from simulated path matrices.

    ``variances`` contains stochastic variance at path time points. For LSV,
    pass ``local_variances=L(t,S)*L(t,S)*V`` at the same points. Integrated
    interval variance is approximated by the trapezoidal rule. Continuous
    monitoring applies conditional Brownian-bridge survival weighting; discrete
    monitoring checks simulated endpoints only.
    """
    geometry = BarrierType(barrier_type)
    payoff = BarrierPayoff(payoff_type)
    monitoring_mode = Monitoring(monitoring)
    spot_paths = np.asarray(spots, dtype=float)
    variance_paths = np.asarray(variances, dtype=float)
    if spot_paths.ndim != 2 or variance_paths.shape != spot_paths.shape:
        raise ValueError("spots and variances must share shape (paths, steps+1)")
    if spot_paths.shape[0] < 2 or spot_paths.shape[1] < 2:
        raise ValueError("at least two paths and one time step are required")
    if np.any(~np.isfinite(spot_paths)) or np.any(spot_paths <= 0.0):
        raise ValueError("spot paths must be finite and positive")
    instantaneous = (
        variance_paths
        if local_variances is None
        else np.asarray(local_variances, dtype=float)
    )
    if instantaneous.shape != spot_paths.shape:
        raise ValueError("local_variances must match the spot path matrix")
    if np.any(~np.isfinite(instantaneous)) or np.any(instantaneous < 0.0):
        raise ValueError(
            "instantaneous local variances must be finite and non-negative"
        )
    if not np.isfinite(maturity) or maturity <= 0.0:
        raise ValueError("maturity must be finite and positive")
    steps = spot_paths.shape[1] - 1
    dt = maturity / steps
    interval_variances = 0.5 * (instantaneous[:, :-1] + instantaneous[:, 1:]) * dt
    lower = 0.0 if lower_barrier is None else lower_barrier
    upper = 0.0 if upper_barrier is None else upper_barrier
    raw = lsv_cpp.price_barrier(
        spot_paths,
        interval_variances,
        maturity,
        rate,
        payout,
        geometry.value,
        payoff.value,
        monitoring_mode.value,
        lower,
        upper,
        seed,
        image_terms,
    )
    return BarrierPrice(
        float(raw["estimate"]),
        float(raw["standard_error"]),
        (
            float(raw["confidence_interval_low"]),
            float(raw["confidence_interval_high"]),
        ),
        int(raw["seed"]),
        int(raw["paths"]),
        int(raw["steps"]),
        monitoring_mode,
    )


def price_dnt(
    spots: ArrayLike,
    variances: ArrayLike,
    maturity: float,
    rate: float,
    payout: float,
    lower_barrier: float,
    upper_barrier: float,
    *,
    monitoring: Monitoring | str = Monitoring.CONTINUOUS,
    local_variances: ArrayLike | None = None,
    seed: int = 0,
) -> BarrierPrice:
    """Price a maturity-paid Double No-Touch claim."""
    return price_barrier(
        spots,
        variances,
        maturity,
        rate,
        payout,
        BarrierType.DOUBLE,
        BarrierPayoff.NO_TOUCH,
        monitoring,
        lower_barrier=lower_barrier,
        upper_barrier=upper_barrier,
        local_variances=local_variances,
        seed=seed,
    )
